Firestore/core/src/firebase/firestore/local/local_serializer.cc

```diff
@@ -208,6 +211,123 @@ std::unique_ptr<model::NoDocument> LocalSerializer::DecodeNoDocument(
                                               version);
 }
 
+util::Status LocalSerializer::EncodeQueryData(
+    const QueryData& query_data, std::vector<uint8_t>* out_bytes) const {
+  HARD_ASSERT(QueryPurpose::kListen == query_data.purpose(),
+              "Only queries with purpose %s may be stored, got %s",
+              QueryPurpose::kListen, query_data.purpose());
+  Writer writer = Writer::Wrap(out_bytes);
+  EncodeQueryData(&writer, query_data);
+  return writer.status();
+}
+
+void LocalSerializer::EncodeQueryData(Writer* writer,
+                                      const QueryData& query_data) const {
+  if (!writer->status().ok()) return;
+
+  writer->WriteTag({PB_WT_VARINT, firestore_client_Target_target_id_tag});
+  writer->WriteInteger(query_data.target_id());
+
+  writer->WriteTag(
+      {PB_WT_STRING, firestore_client_Target_snapshot_version_tag});
+  writer->WriteNestedMessage([&](Writer* writer) {
+    rpc_serializer_.EncodeTimestamp(writer,
+                                    query_data.snapshot_version().timestamp());
+  });
+
+  writer->WriteTag({PB_WT_STRING, firestore_client_Target_resume_token_tag});
+  writer->WriteBytes(query_data.resume_token());
+
+  const Query& query = query_data.query();
+  if (query.IsDocumentQuery()) {
+    // TODO(rsgowman): Implement. Probably like this (once EncodeDocumentsTarget
+    // exists):
+    /*
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_documents_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeDocumentsTarget(writer, query);
+    });
+    */
+    abort();
+  } else {
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_query_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeQueryTarget(writer, query);
+    });
+  }
+}
+
+util::StatusOr<QueryData> LocalSerializer::DecodeQueryData(
+    const uint8_t* bytes, size_t length) const {
+  Reader reader = Reader::Wrap(bytes, length);
+  QueryData query_data = DecodeQueryData(&reader);
+  if (reader.status().ok()) {
+    return query_data;
+  } else {
+    return reader.status();
+  }
+}
+
+QueryData LocalSerializer::DecodeQueryData(Reader* reader) const {
+  if (!reader->status().ok()) return QueryData::Invalid();
+
+  int64_t target_id = 0;
+  SnapshotVersion version = SnapshotVersion::None();
+  std::vector<uint8_t> resume_token;
+  Query query = Query::Invalid();
+
+  while (reader->bytes_left()) {
+    Tag tag = reader->ReadTag();
+    if (!reader->status().ok()) return QueryData::Invalid();
```

Comment:
Is this really required?

What if ReadTag set a field_number that's guaranteed not to match anything on failure?

Then our `default` case would kick in, we'd try to skip a field, but that would no-op because there's already an error, and then we'd loop again and break.

Reply:
Thought about this some more and I think we have an opportunity for further simplification if we embed the current tag in the reader. ReadTag could assign the current tag and return just the tag number.

That would make this code:
```c++
while (reader->bytes_left()) {
  switch (reader->ReadTag()) {
    // ...
```

Reply:
The intent of these comments are that we should heavily bias the `Reader` API toward minimizing the amount of code we have to write to implement these oneof blocks. Since every while loop is going to check status.ok and bytes_left we should have an method that just does both.

We can either define bytes_left() to return 0 if !status.ok() or define some alternative accessor that takes both into account. I imagine that once we defined the alternative the number of uses of bytes_left with its current definition would drop to zero (which is why I propose repurposing it).